src/realm/util/file.cpp

```diff
@@ -985,8 +997,70 @@ bool File::lock(bool exclusive, bool non_blocking)
         return false;
     throw std::system_error(err, std::system_category(), "LockFileEx() failed");
 
-#else // BSD / Linux flock()
+#else
+#ifdef FILELOCK_EMULATION
+    // If we already have any form of lock, release it before trying to acquire a new
+    if (m_has_exclusive_lock || m_has_shared_lock)
+        unlock();
+    // First obtain an exclusive lock on the file proper
+    int operation = LOCK_EX;
+    if (non_blocking)
+        operation |= LOCK_NB;
+    int status;
+    do {
+        status = flock(m_fd, operation);
+    } while (status != 0 && errno == EINTR);
+    if (status != 0 && errno == EWOULDBLOCK)
+        return false;
+    if (status != 0)
+        throw std::system_error(errno, std::system_category(), "flock() failed");
+    // now use a named pipe to emulate locking in conjunction with using exclusive lock
+    // on the file proper.
+    // exclusively locked: we can't sucessfully write to the pipe.
+    //                     AND we continue to hold the exclusive lock.
+    //                     (unlock must lift the exclusive lock).
+    // shared locked: we CAN succesfully write to the pipe. We open the pipe for reading
+    //                before releasing the exclusive lock.
+    //                (unlock must close the pipe for reading)
+    REALM_ASSERT_RELEASE(m_pipe_fd == -1);
+    // TODO create the pipe if it doesn't exist.
+    if (m_fifo_path.empty())
+        m_fifo_path = m_path + ".fifo";
+    status = mkfifo(m_fifo_path.c_str(), 0666);
+    REALM_ASSERT(status == 0 || (status == -1 && errno == EEXIST));
+    if (exclusive) {
+        // check if any shared locks are already taken by trying to open the pipe for writing
+        // shared locks are indicated by one or more readers already having opened the pipe
+        int fd = ::open(m_fifo_path.c_str(), O_WRONLY | O_NONBLOCK);
+        if (fd == -1 && errno != ENXIO)
+            throw std::system_error(errno, std::system_category(), "opening fifo for writing failed");
```

Comment:
"opening lock fifo for writing failed" (and similarly for reading below) would make it a little clearer what went wrong if this ever does.